python-package/lightgbm/basic.py

```diff
@@ -637,6 +638,58 @@ def inner_predict(mat, num_iteration, predict_type, preds=None):
         else:
             return inner_predict(mat, num_iteration, predict_type)
 
+    def __convert_ctypes_to_numpy(self, ctypes_array, ctypes_array_len, array_type):
+        """Convert the ctypes array to a numpy array, note memory will still need to be managed."""
+        array_size = np.dtype(array_type).itemsize * ctypes_array_len
+        if sys.version_info.major >= 3:
+            bfm = ctypes.pythonapi.PyMemoryView_FromMemory
+            bfm.restype = ctypes.py_object
+            bfm.argtypes = (ctypes.c_void_p, ctypes.c_int, ctypes.c_int)
+            PyBUF_READ = 0x100
+            py_buffer = bfm(ctypes_array, array_size, PyBUF_READ)
+        else:
+            bfm = ctypes.pythonapi.PyBuffer_FromMemory
+            bfm.restype = ctypes.py_object
+            bfm.argtypes = (ctypes.c_void_p, ctypes.c_int)
+            py_buffer = bfm(ctypes_array, array_size)
+        return np.frombuffer(py_buffer, array_type)
```

Comment:
difference of this function to https://github.com/microsoft/LightGBM/blob/f155379c1fe4e53d5d1c7fd195987a8d3aec4331/python-package/lightgbm/basic.py#L88-L109 ?

Reply:
great question, I wasn't sure myself at first what the difference is between np.frombuffer and np.fromiter, but based on this:
https://stackoverflow.com/questions/58091879/whats-the-difference-between-frombuffer-and-fromiter-in-numpywhy-and-when-to-us
and docs here:
https://numpy.org/doc/stable/reference/generated/numpy.frombuffer.html
https://numpy.org/doc/stable/reference/generated/numpy.fromiter.html

it looks like frombuffer uses the existing ctypes array whereas fromiter allocates a new array, so I think it's better to use frombuffer in order to reduce memory usage and data copying.

```
 numpy.frombuffer(buffer, dtype=float, count=-1, offset=0)
    Interpret a buffer as a 1-dimensional array.
```
```
 numpy.fromiter(iterable, dtype, count=-1)
    Create a new 1-dimensional array from an iterable object.
```

Reply:
great ! also ping @StrikerRUS , for the performance, should we avoid the copy as possible?

Reply:
@imatiach-msft however, as these memories are in c++, you need a copy after that. like line 696 and 698.

Reply:
@guolinke good point, I am doing a copy regardless afterward, since the array was allocated from native c++ and numpy can't track it.  Do you think it would be better to try to use the existing methods and do the copy then, to avoid the copies on lines 696 and 698, or should I leave the code as-is?

Reply:
@guolinke 
> also ping @StrikerRUS , for the performance, should we avoid the copy as possible?

yeah, absolutely! I remember we had some reports about that conversion from `pandas` used to double the memory usage and we kind of fixed it.

Reply:
@imatiach-msft if there are no performance gain, we think we should keep the code simple, and reuse the existing methods as possible.

Reply:
done! changed the code to use the existing methods that utilize fromiter instead of frombuffer, re-ran the test and it still seems to be working